src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs

```diff
@@ -237,20 +237,27 @@ private BoundInterpolatedString BindUnconvertedInterpolatedStringToString(BoundU
             // We need to do the determination of 1, 2, 3, or 4/5 up front, rather than in lowering, as it affects diagnostics (ref structs not being
             // able to be used, for example). However, between 4 and 5, we don't need to know at this point, so that logic is deferred for lowering.
 
+            // Don't apply optimizations in expression trees because we can't hide them.
+            // We have to disguise them as string.Format calls without modifying parts.
+
+            var canHideOptimization = !InExpressionTree && unconvertedInterpolatedString.Type is { SpecialType: SpecialType.System_String };
+            var literalMergedParts = canHideOptimization
+                ? MergeStringConstLikeInPart(unconvertedInterpolatedString.Parts, unconvertedInterpolatedString.Syntax, diagnostics)
```

Comment:
We cannot do this optimization in initial binding, as it will affect observable results from our public APIs. It must be done in the LocalRewriter. You _can_ update the BoundNodes if you need to calculate some information during initial binding to make it easier to do the optimization step, but I expect the shape of the bound tree (and the IOperation tree by extension) to reflect the written code, not the optimized code.

Reply:
@333fred I thought that I could make all things done jusy by modifying the LocalRewriter, but actually Binder *does* generate calls for `AppendLiteral` or `AppendFormat` in  DefaultStringInterpolationHandler, or `string.Concat`.

https://github.com/dotnet/roslyn/blob/7809895241dc3931ca12054a08761bf34e6f6fae/src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringInterpolation.cs#L283-L286

https://github.com/dotnet/roslyn/blob/7809895241dc3931ca12054a08761bf34e6f6fae/src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringInterpolation.cs#L77

Should we add a new property named `optimizedParts` in `InterpolatedStringHandlerData`?

Reply:
Without processing in Binder, we cannot transform `$"{"a"}b{"c"}d{'e'}{stringVariable}f{"g"}"` into `string.Concat("abcde", stringVariable, "fg")`.

Reply:
I expect that you can extract out components from these calls in the LocalRewriter and combine them, yes. Specifically for interpolations converted to `string`, look for `DefaultInterpolatedStringHandler.AppendFormatted(string)` calls, and see if you can combine it with adjacent `AppendLiteral` calls. You could be able to do similar for `AppendFormatted<char>(char)`.

Reply:
I am trying, but I have no idea about how to convert BoundCall for AppendFormatted to that for AppendLiteral in a sub method in `LowerPartsToString`.

Reply:
> I am trying, but I have no idea about how to convert BoundCall for AppendFormatted to that for AppendLiteral in a sub method in `LowerPartsToString`. If `_factory.InstanceCall` were correct, I do not know what the correct `receiver` is.

The receiver would be the same receiver as the original `AppendFormatted` call.

> P.S. Is `_factory.StaticCall(data.BuilderType, BoundInterpolatedString.AppendLiteralMethod, _factory.StringLiteral("joinned literal"))` correct? It look like its syntax tree is wrong, though.

No, `AppendLiteral` is not a static method.

Reply:
@333fred I managed to use `_factory.InstanceCall` by using `.ReceiverOpt`.